### PR TITLE
Export socket options related macros

### PR DESCRIPTION
**Pull request by mexus**
In this PR I've added docs to that macros and exported them.
I've also made related structs/traits to be public as well.
Hope I didn't forget anything :)

Reference issue: #577.

### CODE REVIEW

**Comment by mexus** [comment]
Is there something wrong with the request?

**Comment by Susurrus** [comment]
I'd be more interested in exposing a proper `SO_ATTACH_FILTER` wrapper rather than exporting all of the macros instead. I'd think that's also be better for downstream as then we own that code rather than them. Was there a reason you took this second approach? If it's feasible to add this wrapper I'd like to add that and only expose this macros as a last resort.

**Comment by mexus** [comment]
Well, the reason was not to force the users to make a pull request every time they need a socket option that is not implemented yet :) But it will be good for the crate, right? And eventually all the options, at least the common ones, will be implemented here.

But nevertheless, I've got your point so lets close this one and I'll just make a SO_ATTACH_FILTER implementation.

**Comment by Susurrus** [comment]
Yes, I'd think it'd overall be better for the crate and our downstream users if we implemented everything in `nix` so that everyone could benefit.

Additionally, I would actually like to merge your documentation changes here, just not the part that exports the macros. Would you be willing to revise this PR in that fashion and we'll merge it?

**Comment by mexus** [comment]
Sure, np.

**Comment by mexus** [comment]
Although the examples will have to be marked with `no_run` :( But I guess they won't be changed much in the future.

File: src/sys/socket/sockopt.rs [hunk 1]
```diff
@@ -25,6 +47,28 @@ macro_rules! setsockopt_impl {
...
+///
+/// This macro aims to help implementing `GetSockOpt` for different socket options that accept
+/// different kinds of data to be use with `getsockopt`.
+///
+/// Instead of using this macro directly consider using `sockopt_impl!`, especially if the option
+/// you are implementing represents a simple type.
+///
+/// # Arguments
+///
+/// 1. Name of the type you want to implement `GetSockOpt` for.
```
Comment (Susurrus) [inline src/sys/socket/sockopt.rs]: Please use bullet points here `*` instead of a numbered list.

File: src/sys/socket/sockopt.rs [hunk 2]
```diff
@@ -25,6 +47,28 @@ macro_rules! setsockopt_impl {
...
+/// 1. Socket layer, or a `protocol level`: could be *raw sockets* (`lic::SOL_SOCKET`),  *ip
+///    protocol* (libc::IPPROTO_IP), *tcp protocol* (`libc::IPPROTO_TCP`),  and more. Please refere
```
Comment (Susurrus) [inline src/sys/socket/sockopt.rs]: "refere" -> "refer"

File: src/sys/socket/sockopt.rs [hunk 3]
```diff
@@ -6,6 +6,28 @@ use libc::{self, c_int, uint8_t, c_void, socklen_t};
...
+///
+/// This macro aims to help implementing `SetSockOpt` for different socket options that accept
+/// different kinds of data to be used with `setsockopt`.
+///
+/// Instead of using this macro directly consider using `sockopt_impl!`, especially if the option
+/// you are implementing represents a simple type.
+///
+/// # Arguments
+///
+/// 1. `$name:ident`: name of the type you want to implement `SetSockOpt` for.
```
Comment (Susurrus) [inline src/sys/socket/sockopt.rs]: Change these to bullet points using `*` instead.

File: src/sys/socket/sockopt.rs [hunk 4]
```diff
@@ -6,6 +6,28 @@ use libc::{self, c_int, uint8_t, c_void, socklen_t};
...
+/// 1. `$level:path` : socket layer, or a `protocol level`: could be *raw sockets*
+///    (`lic::SOL_SOCKET`), *ip protocol* (libc::IPPROTO_IP), *tcp protocol* (`libc::IPPROTO_TCP`),
+///    and more. Please refere to your system manual for more options. Will be passed as the second
```
Comment (Susurrus) [inline src/sys/socket/sockopt.rs]: "refere" -> "refer"

File: src/sys/socket/sockopt.rs [hunk 5]
```diff
@@ -46,7 +90,31 @@ macro_rules! getsockopt_impl {
...
 
-// Helper to generate the sockopt accessors
+/// Helper to generate the sockopt accessors. See
+/// [`::sys::socket::GetSockOpt`](sys/socket/trait.GetSockOpt.html) and
+/// [`::sys::socket::SetSockOpt`](sys/socket/trait.SetSockOpt.html).
+///
+/// This macro aims to help implementing `GetSockOpt` and `SetSockOpt` for different socket options
+/// that accept different kinds of data to be use with `getsockopt` and `setsockopt` respectively.
+///
+/// Basically this macro wraps up [`getsockopt_impl!`](macro.getsockopt_impl.html) and
```
Comment (Susurrus) [inline src/sys/socket/sockopt.rs]: "wraps up " -> "wraps up the "

File: src/sys/socket/sockopt.rs [hunk 6]
```diff
@@ -46,7 +90,31 @@ macro_rules! getsockopt_impl {
...
+/// [`setsockopt_impl!`](macro.setsockopt_impl.html) macros.
+///
+/// # Arguments
+///
+/// 1. `GetOnly`, `SetOnly` or `Both`: whether you want to implement only getter, only setter or
```
Comment (Susurrus) [inline src/sys/socket/sockopt.rs]: Change to bullet points using `*` instead

File: src/sys/socket/sockopt.rs [hunk 7]
```diff
@@ -46,7 +90,31 @@ macro_rules! getsockopt_impl {
...
+///    both of them.
+/// 1. `$name:ident`: name of type `GetSockOpt`/`SetSockOpt` will be implemented for.
+/// 1. `$level:path` : socket layer, or a `protocol level`: could be *raw sockets*
+///    (`lic::SOL_SOCKET`), *ip protocol* (libc::IPPROTO_IP), *tcp protocol* (`libc::IPPROTO_TCP`),
+///    and more. Please refere to your system manual for more options. Will be passed as the second
```
Comment (Susurrus) [inline src/sys/socket/sockopt.rs]: "refere" -> "refer"

**Comment by mexus** [comment]
Hope that fits :)

**Comment by Susurrus** [comment]
Awesome, thanks!

bors r+

**Comment by tomwhoiscontrary** [comment]
Do i understand correctly that the PR as merged documented the internal functions and types, but did not expose them? So there's still no way to define custom socket options?

**Comment by mexus** [comment]
I'd say you've got right the part about the docs, but there is a way to at least use custom options: `setsockopt`/`getsockopt` accept objects that implement `SetSockOpt`/`GetSockOpt` which are public traits, so it shouldn't big a big deal. Yes, the `set` and `get` methods of the traits are hidden from the docs (for the reasons unknown to me), but still.

If by "custom" you mean something actually common (i.e. from the libc) but which hasn't been implemented yet inside the *nix* crate, it might be a good idea to either post an issue or make a PR with the desired option. If it's something really custom, then I guess implementing the traits once shouldn't be an issue..

Having said that I want to add that if I was a maintainer I'd definitely make the macro public with some comments like "please consider sending a PR if you're implementing something common!" :)

**Comment by tomwhoiscontrary** [comment]
Aha, that is probably enough.

The specific situation we were thinking about was for some funky hardware where the vendor-supplied TCP/IP stack has a few extra socket options to configure it. I don't think it would make sense to put those into Nix.

We would like to see [SO_TIMESTAMPING](https://www.kernel.org/doc/Documentation/networking/timestamping.txt) in Nix, though. That would also need some machinery to decode the timestamps returned in the cmsg ancillary data, which is not terrifically complicated. We'll see if we can get a PR in for this ...

**Pull request merged**